src/k2/module/k23si/TxnManager.cpp

```diff
@@ -192,247 +199,482 @@ TxnRecord& TxnManager::_createRecord(dto::TxnId txnId) {
+            if (!flushStatus.is2xxOK()) {
+                return seastar::make_ready_future<Status>(std::move(flushStatus));
+            }
+            return std::move(finfut);
+        });
+    }
+    _addBgTask(rec, [finfut=std::move(finfut)] () mutable { return finfut.discard_result();});
+    return seastar::make_ready_future<Status>(dto::K23SIStatus::OK);
+}
 
-    return _persistence->flush().then([this, &rec] (auto&& flushStatus) {
-        if (!flushStatus.is2xxOK()) {
-            return seastar::make_exception_future(ServerError("persistence flush failed"));
-        }
+seastar::future<Status> TxnManager::_commit(TxnRecord& rec) {
+    K2LOG_D(log::skvsvr, "Setting status to Commit for {}", rec);
+    rec.state = dto::TxnRecordState::Committed;
+    return _endHelper(rec);
+}
 
-        auto timeout = (10s + _config.writeTimeout() * rec.writeKeys.size()) / _config.finalizeBatchSize();
+seastar::future<Status> TxnManager::_abort(TxnRecord& rec) {
+    K2LOG_D(log::skvsvr, "Setting status to Abort for {}", rec);
+    rec.state = dto::TxnRecordState::Aborted;
+    return _endHelper(rec);
+}
 
-        if (rec.syncFinalize) {
-            // append to WAL
-            _persistence->append(rec);
+seastar::future<Status> TxnManager::_endHelper(TxnRecord& rec) {
+    K2LOG_D(log::skvsvr, "Processing END for {}", rec);
 
-            return _finalizeTransaction(rec, FastDeadline(timeout));
-        }
-        else {
-            // enqueue in background tasks
-            rec.bgTaskFut = rec.bgTaskFut
-                .then([&rec] {
-                    return seastar::sleep(rec.timeToFinalize);
-                })
-                .then([this, &rec, timeout]() {
-                    return _finalizeTransaction(rec, FastDeadline(timeout));
+    auto timeout = (10s + _config.writeTimeout() * rec.writeKeys.size()) / _config.finalizeBatchSize();
+
+    if (rec.syncFinalize) {
+        return _finalizeTransaction(rec, FastDeadline(timeout));
+    }
+
+    // we're doing async finalize. enqueue in background tasks
+    _addBgTask(rec,
+        [this, &rec, timeout] {
+            return seastar::sleep(rec.timeToFinalize)
+                .then([this, &rec, timeout] {
+                    return _finalizeTransaction(rec, FastDeadline(timeout)).discard_result();
                 });
-            // append to WAL
-            _persistence->append(rec);
-            return seastar::make_ready_future();
-        }
-    });
+        });
+
+    return seastar::make_ready_future<Status>(dto::K23SIStatus::OK);
 }
 
-seastar::future<> TxnManager::_deleted(TxnRecord& rec) {
-    K2LOG_D(log::skvsvr, "Setting status to deleted for {}", rec);
+seastar::future<Status> TxnManager::_finalizedPIP(TxnRecord& rec) {
+    K2LOG_D(log::skvsvr, "Setting status to FinalizedPIP for {}", rec);
     // set state
-    rec.state = dto::TxnRecordState::Deleted;
+    rec.state = dto::TxnRecordState::FinalizedPIP;
     // manage hb expiry
     rec.unlinkHB(_hblist);
```

Comment:
These two unlink looks like redundant. Maybe we can assert here that this tr doesn't exist in these two lists for debugging purpose.

Reply:
cleaned up